homeassistant/components/iotty/cover.py

```diff
@@ -0,0 +1,218 @@
+from typing import Any, cast
+
+from iottycloud.device import Device
+from iottycloud.shutter import Shutter, ShutterState
+from iottycloud.verbs import SH_DEVICE_TYPE_UID
+
+from homeassistant.components.cover import (
+    ATTR_POSITION,
+    CoverDeviceClass,
+    CoverEntity,
+    CoverEntityFeature,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from . import IottyConfigEntry
+from .api import IottyProxy
+from .const import DOMAIN
+from .coordinator import IottyDataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: IottyConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Activate the iotty Shutter component."""
+    _LOGGER.debug("Setup COVER entry id is %s", config_entry.entry_id)
+
+    coordinator = config_entry.runtime_data.coordinator
+    entities = [
+        IottyShutter(
+            coordinator=coordinator, iotty_cloud=coordinator.iotty, iotty_device=d
+        )
+        for d in coordinator.data.devices
+        if d.device_type == SH_DEVICE_TYPE_UID
+        if (isinstance(d, Shutter))
+    ]
+    _LOGGER.debug("Found %d Shutters", len(entities))
+
+    async_add_entities(entities)
+
+    known_devices: set = config_entry.runtime_data.known_devices
+    for known_device in coordinator.data.devices:
+        if known_device.device_type == SH_DEVICE_TYPE_UID:
+            known_devices.add(known_device)
+
+    @callback
+    def async_update_data() -> None:
+        """Handle updated data from the API endpoint."""
+        if not coordinator.last_update_success:
+            return
+
+        devices = coordinator.data.devices
+        entities = []
+        known_devices: set = config_entry.runtime_data.known_devices
+
+        # Add entities for devices which we've not yet seen
+        for device in devices:
+            if (
+                any(d.device_id == device.device_id for d in known_devices)
+                or device.device_type != SH_DEVICE_TYPE_UID
+            ):
+                continue
+
+            iotty_entity = IottyShutter(
+                coordinator=coordinator,
+                iotty_cloud=coordinator.iotty,
+                iotty_device=Shutter(
+                    device.device_id,
+                    device.serial_number,
+                    device.device_type,
+                    device.device_name,
+                ),
+            )
+
+            entities.extend([iotty_entity])
+            known_devices.add(device)
+
+        async_add_entities(entities)
+
+    # Add a subscriber to the coordinator to discover new devices
+    coordinator.async_add_listener(async_update_data)
+
+
+class IottyShutter(CoverEntity, CoordinatorEntity[IottyDataUpdateCoordinator]):
```

Comment:
Consider a preliminary PR where you create a shared base entity

Reply:
Hi, what do you mean exactly? You'd like an extra abstraction that both the IottySwitch and the IottyShutter inherit from?
I think that's an interesting approach, but since the guidelines on the docs are not really clear about what should and what should not be done I preferred going for the simple approach

Reply:
It's not something in the docs but in the code, check AirGradient for an example

Reply:
Looks clear enough to me based on the example you are proposing. Is it mandatory to do it before this PR? I'd rather do it afterwards, especially because with two entities it would be easier to refactor later and find the correct common denominator for the base class, otherwise if I did it starting from the current dev version I would just add a layer of abstraction that references itself only.

Reply:
You can do it in this PR or a separate one upfront

Reply:
I did it here!